core/src/km_core_action_api.cpp

```diff
@@ -33,7 +33,22 @@ km_core_actions const * km_core_state_get_actions(
     return nullptr;
   }
 
-  return action_item_list_to_actions_object(action_items);
+  km_core_actions * result = action_item_list_to_actions_object(action_items);
+
+  if(state->processor().supports_normalization()) {
+    // Normalize to NFC for those keyboard processors that support it
+    if(!actions_normalize(km_core_state_context(state), km_core_state_app_context(state), result)) {
+      km_core_actions_dispose(result);
+      return nullptr;
+    }
+  } else {
+    // For all other keyboard processors, we just copy the cached_context to the app_context
+    if(!actions_update_app_context(km_core_state_context(state), km_core_state_app_context(state))) {
```

Comment:
Perhaps we should add an `_nfu` to this to clarify this is for non-normalized use only?

Reply:
I think that would make sense. And the comment on `actions_update_app_context_nfu` should mention that this will be used for processors that don't support normalization.